lib/screens/editor/versions.dart

```diff
@@ -0,0 +1,398 @@
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:every_door/private.dart';
+import 'package:every_door/providers/osm_api.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+import 'package:http/http.dart' as http;
+import 'package:intl/intl.dart';
+import 'package:url_launcher/url_launcher.dart';
+
+class TagChange {
+  final String? oldValue;
+  final String? newValue;
+  bool get noChange => oldValue == newValue;
+  bool get removed => newValue == null;
+  bool get added => oldValue == null;
+  bool get updated => !added && !removed && oldValue != newValue;
+
+  const TagChange(this.oldValue, this.newValue);
+
+  Color? getColor() {
+    if (noChange) return null;
+    if (added) return Colors.green.shade100;
+    if (updated) return Colors.amber.shade200;
+    if (removed) return Colors.red.shade100;
+
+    // should never be reached
+    return null;
+  }
+
+  @override
+  String toString() {
+    return "was: $oldValue, now: $newValue";
+  }
+}
+
+class Version {
+  final int number;
+  final String user;
+  final DateTime timestamp;
+  final int changeset; // used to fetch changeset comment
+  final Map<String, String> tags;
+  final bool isLocal;
+  String? comment; // fetched separately
+
+  /// stores the tags changes between previous versions and this versions.
+  /// uses a `SplayTreeMap` so that keys are sorted alphabetically
+  Map<String, TagChange> tagChanges = SplayTreeMap();
+
+  // don't need to show diff when no tags have changed
+  bool get noTagChange => tagChanges.entries.every((e) => e.value.noChange);
+
+  factory Version.fromJson(Map<String, dynamic> data) {
+    final number = data['version'] as int;
+    // some old changesets were created anonymously
+    final user = (data['user'] ?? 'anonymous') as String;
+    final timestamp = DateTime.parse(data['timestamp'] as String);
+    final changeset = data['changeset'] as int;
+    final tags =
+        (data['tags'] ?? {}).cast<String, String>() as Map<String, String>;
+
+    return Version(
+      number: number,
+      user: user,
+      timestamp: timestamp,
+      changeset: changeset,
+      tags: tags,
+    );
+  }
+
+  @override
+  String toString() {
+    return 'v$number: user: $user, timestamp: $timestamp, changeset: $changeset, comment: $comment, tags: $tags updated: $tagChanges';
+  }
+
+  Version({
```

Comment:
Please move constructors to be the first methods in each class.

Reply:
Fixed here 2e563be166f59791df1185aa2e656523515cf28e